core/src/main/java/io/grpc/internal/GzipInflatingBuffer.java

```diff
@@ -0,0 +1,514 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.io.Closeable;
+import java.util.zip.CRC32;
+import java.util.zip.DataFormatException;
+import java.util.zip.Inflater;
+import java.util.zip.ZipException;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Processes gzip streams, delegating to {@link Inflater} to perform on-demand inflation of the
+ * deflated blocks. Like {@link java.util.zip.GZIPInputStream}, this handles concatenated gzip
+ * streams. Unlike {@link java.util.zip.GZIPInputStream}, this allows for incremental processing of
+ * gzip streams, allowing data to be inflated as it arrives over the wire.
+ *
+ * <p>This also frees the inflate context when the end of a gzip stream is reached without another
+ * concatenated stream available to inflate.
+ */
+@NotThreadSafe
+class GzipInflatingBuffer implements Closeable {
+
+  private static final int INFLATE_BUFFER_SIZE = 512;
+  private static final int MAX_OUTPUT_BUFFER_SIZE = 1024 * 4;
+  private static final int UNSIGNED_SHORT_SIZE = 2;
+
+  private static final int GZIP_MAGIC = 0x8b1f;
+
+  private static final int GZIP_HEADER_MIN_SIZE = 10;
+  private static final int GZIP_TRAILER_SIZE = 8;
+
+  private static final int HEADER_CRC_FLAG = 2;
+  private static final int HEADER_EXTRA_FLAG = 4;
+  private static final int HEADER_NAME_FLAG = 8;
+  private static final int HEADER_COMMENT_FLAG = 16;
+
+  /**
+   * Reads gzip header and trailer bytes from the inflater's buffer (if bytes beyond the inflate
+   * block were given to the inflater) and then from {@code gzippedData}, and handles updating the
+   * CRC and the count of gzipped bytes consumed.
+   */
+  private class GzipMetadataReader {
+
+    /**
+     * Returns the next unsigned byte, adding it the CRC and incrementing {@code bytesConsumed}.
+     *
+     * <p>It is the responsibility of the caller to verify and reset the CRC as needed, as well as
+     * caching the current CRC value when necessary before invoking this method.
+     */
+    private int readUnsignedByte() {
+      int bytesRemainingInInflater = inflaterInputEnd - inflaterInputStart;
+      int b;
+      if (bytesRemainingInInflater > 0) {
+        b = inflaterInput[inflaterInputStart] & 0xFF;
+        if (inflater != null) {
+          inflater.reset();
+          inflater.setInput(inflaterInput, inflaterInputStart + 1, bytesRemainingInInflater - 1);
```

Comment:
Leaving the `inflaterBuf` in the `inflater` even when not inflating is weird to me. Why wouldn't you just have the inflater ready when you are inflating... That would remove the `setInput()` noise here.